### PR TITLE
Event handlers on document or window should not be called with shadow  element as target (#2153)

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

**Comment by AlexKamaev** [comment]
@testcafe-build-bot \retest

**Comment by AlexKamaev** [comment]
@testcafe-build-bot \retest

File: src/client/ui/select-element.js [hunk 1]
```diff
@@ -36,6 +37,23 @@ function onDocumentMouseDown (e) {
         collapseOptionList();
 }
 
+function onWindowClick (e, dispatched, preventDefault) {
+    const target   = e.target || e.srcElement;
```
Comment (helen-dikareva) [inline src/client/ui/select-element.js]: As far as I know `srcElement` exists only in IE 6-8 so we can get rid of it.

File: src/client/ui/select-element.js [hunk 2]
```diff
@@ -36,6 +37,23 @@ function onDocumentMouseDown (e) {
...
+    const optionIndex = arrayUtils.indexOf(options, target);
+
+    if (optionIndex < 0)
+        return;
+
+    preventDefault();
+
+    const isDisabled = target.className.indexOf(DISABLED_CLASS) > -1;
```
Comment (helen-dikareva) [inline src/client/ui/select-element.js]: can we just check here `target.isDisabled`?

File: src/client/ui/select-element.js [hunk 3]
```diff
@@ -36,6 +37,23 @@ function onDocumentMouseDown (e) {
...
+
+    if (isDisabled && browserUtils.isWebKit)
+        return;
```
Comment (helen-dikareva) [inline src/client/ui/select-element.js]: Later we return `false` in this case

File: test/functional/fixtures/regression/gh-2153/pages/index.html [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Title</title>
+    <script type="text/javascript">
+        window.counter = 0;
+
+        function onClick (e) {
+            var currentTarget = e.currentTarget;
+            var logger = document.getElementById('logger');
```
Comment (helen-dikareva) [inline test/functional/fixtures/regression/gh-2153/pages/index.html]: we always add line break after a vars declaration:
```js
var/const/let a = 'foo';

//other code
```
OR
```js
a.foo = 'for';

//other code
```

File: test/functional/fixtures/regression/gh-2153/testcafe-fixtures/index.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { Selector } from 'testcafe';
+
+fixture `GH-2153 - Shouldn't call document and window handlers extra times`
+    .page `http://localhost:3000/fixtures/regression/gh-2153/pages/index.html`;
+
+test(`Shouldn't call document and window handlers extra times`, async t => {
+
```
Comment (helen-dikareva) [inline test/functional/fixtures/regression/gh-2153/testcafe-fixtures/index.js]: extra blank line

File: src/client/ui/select-element.js (hunk 2 above)
Comment (AlexKamaev) [reply src/client/ui/select-element.js]: no, it does not work in IE. Probably it is an other bug

File: src/client/ui/select-element.js [hunk 6]
```diff
@@ -78,6 +96,8 @@ function createOption (realOption, parent) {
                                                   realOption.parentElement.disabled;
 
     option.textContent = realOption.text;
+    option.isDisabled  = isOptionDisabled;
```
Comment (AlexKamaev) [inline src/client/ui/select-element.js]: oops, this line should not be here

File: src/client/ui/select-element.js (hunk 3 above)
Comment (AlexKamaev) [reply src/client/ui/select-element.js]: We returned false in onclick handler of shadow option element. Now we handle click on window so we dont need it at all. Events will be prevented in any cases

**Comment by AlexKamaev** [comment]
FPR

File: test/functional/fixtures/regression/gh-2153/pages/index.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Title</title>
+    <script type="text/javascript">
+        window.counter = 0;
+
+        function onClick (e) {
+            var selectEl = document.getElementById('select');
+            var optOne = document.getElementById('one');
```
Comment (helen-dikareva) [inline test/functional/fixtures/regression/gh-2153/pages/index.html]: Test contains click only on the `option` with text `"Two"`.
Let's do not check the `one` option here, e.g.
```js
if (select !== e.target && optTwo !== e.target)
    throw new Error('Shadow element appeared in user event handler');
```

File: test/functional/fixtures/regression/gh-2153/pages/index.html [hunk 8]
```diff
@@ -0,0 +1,31 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Title</title>
+    <script type="text/javascript">
+        window.counter = 0;
```
Comment (helen-dikareva) [inline test/functional/fixtures/regression/gh-2153/pages/index.html]: We don't need it anymore

**Pull request merged**